cmd/dao_tests/account_test.go

```diff
@@ -0,0 +1,164 @@
+	return &models.Account{
+		OrgID:         "2",
+		AccountNumber: sql.NullString{},
+	}
+}
+
+func setupAccount(t *testing.T) (dao.AccountDao, context.Context) {
+	setup()
+	ctx := identity.WithTenant(t, context.Background())
+	accDao, err := dao.GetAccountDao(ctx)
+	if err != nil {
+		panic(err)
+	}
+	return accDao, ctx
+}
+
+func teardownAccount(_ *testing.T) {
+	teardown()
+}
+
+func TestCreateAccount(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	acc := createAccount()
+	err := accDao.Create(ctx, acc)
+	if err != nil {
+		panic(err)
+	}
+	account, err := accDao.GetById(ctx, 2)
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, acc.OrgID, account.OrgID)
+	assert.Equal(t, acc.AccountNumber.String, account.AccountNumber.String)
+	assert.Equal(t, acc.AccountNumber.Valid, account.AccountNumber.Valid)
+}
+
+func TestCreateAccountWithNullAccountNumber(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	acc := createAccountWithNullAccountNumber()
+	err := accDao.Create(ctx, acc)
+	if err != nil {
+		panic(err)
+	}
+	account, err := accDao.GetById(ctx, 2)
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, acc.OrgID, account.OrgID)
+	assert.Equal(t, acc.AccountNumber.String, account.AccountNumber.String)
+	assert.Equal(t, acc.AccountNumber.Valid, account.AccountNumber.Valid)
+}
+
+func TestListAccount(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	accounts, err := accDao.List(ctx, 100, 0)
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, 1, len(accounts))
+}
+
+func TestGetByIdAccount(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	account, err := accDao.GetById(ctx, 1)
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, "1", account.OrgID)
+	assert.Equal(t, "1", account.AccountNumber.String)
+}
+
+func TestGetByAccountNumber(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	account, err := accDao.GetByAccountNumber(ctx, "1")
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, "1", account.OrgID)
+}
+
+func TestGetByOrgId(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	account, err := accDao.GetByOrgId(ctx, "1")
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, int64(1), account.ID)
+	assert.Equal(t, "1", account.AccountNumber.String)
+}
+
+func TestGetOrCreateByIdentityGet(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	account, err := accDao.GetOrCreateByIdentity(ctx, "1", "1")
+	if err != nil {
+		panic(err)
+	}
+
+	assert.Equal(t, "1", account.OrgID)
+	assert.Equal(t, "1", account.AccountNumber.String)
+}
+
+func TestGetOrCreateByIdentityAccountCreate(t *testing.T) {
+	accDao, ctx := setupAccount(t)
+	defer teardownAccount(t)
+	accountsBefore, err := accDao.List(ctx, 100, 0)
+	if err != nil {
+		panic(err)
+	}
+	_, err = accDao.GetOrCreateByIdentity(ctx, "2", "100")
```

Comment:
What this is supposed to return? If it does return an account, I would also assert the result.

Reply:
This should return the account (found if it already exists, created if no such record is in database). GetByOrgId is used here to check if the record was created. We could assert if the return value of this method is correct, but I would say that it does not belong to dao tests, does it?

Reply:
The main purpose of most tests is to execute various code paths, here we could test the path when identity did not exist and the function actually had to construct new record and return a valid model value. I would definitely not throw the value away because GetByOrgId call below tests a different part of the code and I was interested of testing correctness of this one:

https://github.com/RHEnVision/provisioning-backend/blob/51e26b1c3c7ffeab20657d2b6dc50b5adbaecf01/internal/dao/sqlx/account_dao.go#L112-L121

Not big deal, the PR was merged. But asserts typically cost nothing, especially if you do not put long messages into them :-)